### ISSUE TITLE
Use Jenkins for open source build automation

### ISSUE THREAD

**Issue by lesleyannjordan** [hyperledger/composer#3134]
# Abstract
Move to using Jenkins for open source build automation

# Motivation
<Why are we doing this?>
Prerequisite for expanding Hyperledger Composer to other platforms (Linux/ONE, Linux/POWER, Windows)

# Specification
<What are we proposing to do?>
Work with Hyperledger CI team to migrate open source builds to Jenkins.

# Rationale
<Why are we proposing to do it this way?>
Outlined in the motivation section of this document.

# Backwards compatibility
<Who does it break?>

# Test cases
<How are we going to test it?>
Appropriate tests should be defined and run.

# Implementation
<Who is going to do it?>
The service team will  implement this using the specification outlined above.

# Copyright
This document is licensed under the Creative Commons Attribution 4.0 International Public License.

**Comment by jt-nti** [comment]
See related issue #3149 (build script testing)

**Comment by jt-nti** [comment]
Hyperledger Fabric is already built using Jenkins and there is information in the [ci-management](https://github.com/hyperledger/ci-management) project. For example:

- https://github.com/hyperledger/ci-management/blob/master/docs/fabric_ci_process.md
- https://github.com/hyperledger/ci-management/blob/master/Sandbox_Setup.md

**Comment by rameshthoomu** [comment]
Hi @jt-nti  I am from hyperledger CI team. Barry asked me to configure composer Jenkins jobs in hyperledger jenkins and integrate e2e tests in fabric daily job to test them daily and publish the results.  Discussed about this with @jt-nti and @nklincoln today.  I will create couple of tasks here to track the progress.

**Comment by jt-nti** [comment]
I've booked a slot on the next community call (10am UTC Wednesday 7th February) to kick this off.

A few things I'd like to discuss are...

1. What must a new build achieve that Travis cannot deliver? (Seems to come down to platform support)
2. Exactly what platforms are required? (I'm not aware of any 'build' activities that require additional platforms, however automated testing is currently not happening on an expanded set of platforms)
3. Does Jenkins support the required platforms (I can see [Bamboo has wider platform support than Travis](https://confluence.atlassian.com/display/BAMBOO0602/remote+agent+supervisor) but not sure about Jenkins)

Unfortunately Jenkins is in the issue title so I'm guessing we will end up using it whatever happens, in which case...

4. Can we maintain travis builds in parallel? (If not, contributors will lose the ability to easily build their own forks)
5. Who can grant Composer contributors access to the Hyperledger Jenkins system? (Both to work on this issue and for future operations and updates)
6. How do we get/configure Jenkins build slaves with all the required prereqs to run Composer builds? (For similar builds, I've created a docker image for use on some cloud/swarm thing to run the builds, which seems to work pretty well)
7. How are credentials managed for deployment etc. (I assume there is some separation from other Hyperledger projects)
8. Is use of JJB mandatory? (I would rather avoid it and keep Jenkinsfile build definitions in the Composer projects, which will be more familiar to our contributors and be much easier to maintain and evolve compared to a separate gerit hyplerledger-ci project)

That's probably enough to begin with! If you're unable to make the call, please leave a comment here or get in touch via the composer-contributors Rocket Chat channel. (I'm sure we'll be able to hijack future community calls as well!)

**Comment by ryjones** [comment]
> Does Jenkins support the required platforms (I can see Bamboo has wider platform support than Travis but not sure about Jenkins)

currently, Jenkins supports P, Z, and X. No plans to support Windows or macOS.

> Can we maintain travis builds in parallel? (If not, contributors will lose the ability to easily build their own forks)

yes, if you like

> Who can grant Composer contributors access to the Hyperledger Jenkins system? (Both to work on this issue and for future operations and updates)

helpdesk@hyperledger.org

> How do we get/configure Jenkins build slaves with all the required prereqs to run Composer builds? (For similar builds, I've created a docker image for use on some cloud/swarm thing to run the builds, which seems to work pretty well)

We use packer to build images. See: https://github.com/hyperledger/ci-management/tree/master/packer

> How are credentials managed for deployment etc. (I assume there is some separation from other Hyperledger projects)

they're managed inside Jenkins

> Is use of JJB mandatory? (I would rather avoid it and keep Jenkinsfile build definitions in the Composer projects, which will be more familiar to our contributors and be much easier to maintain and evolve compared to a separate gerit hyplerledger-ci project)

Yes, JJB is required. However, using JJB for minion creation/destruction and log retention doesn't mean you can't vary your build based on files in your project

**Comment by jt-nti** [comment]
Jenkins access requested: https://rt.linuxfoundation.org/SelfService/Display.html?id=52008

**Comment by jt-nti** [comment]
Looks like travis builds are failing in forks at the moment #3320

**Comment by jt-nti** [comment]
Thanks for all the answers @ryjones, although unfortunately that inevitably means more questions :)

What's the issue with Windows and OSX? For example, I was under the impression that Jenkins and Hyperledger Fabric supported Windows? (Travis supports OSX, so maybe we could make use of that if required. ARM/Raspberry Pi support would be lovely too but that's probably just me!)

Packer looks good but I'm unsure what the process would be to make a Composer build machine image available for Jenkins builds. For example...
- Could the Packer templates/scripts/etc. stay in the composer repository?
- Are the OpenStack env vars something that would be specific to the Composer project, or are they shared with all Hyperledger projects?
- Is there a CI process for building the images and publishing them somewhere that the Jenkins builds expect to find them?

Thanks

**Comment by ryjones** [comment]
@jt-nti could I ask you to send me an email please with the email address of your LFID? thank you rjones@linuxfoundation.org

**Comment by ryjones** [comment]
> What's the issue with Windows and OSX? For example, I was under the impression that Jenkins and Hyperledger Fabric supported Windows? (Travis supports OSX, so maybe we could make use of that if required. ARM/Raspberry Pi support would be lovely too but that's probably just me!)

We would need to rent macOS servers somewhere to support macOS builds. The expense is high, and would need to be approved - on the order of a few thousand a month. there's more to it than renting a Mac Mini somewhere.

To support Windows, we would need to use something like Azure. Again, the expense is pretty high.

> Packer looks good but I'm unsure what the process would be to make a Composer build machine image available for Jenkins builds. For example...

We use one image to support everything in Jenkins.

> Could the Packer templates/scripts/etc. stay in the composer repository?

not easily. We're pretty quick to merge changes to the ci-management repo.

> Are the OpenStack env vars something that would be specific to the Composer project, or are they shared with all Hyperledger projects?

shared

> Is there a CI process for building the images and publishing them somewhere that the Jenkins builds expect to find them?

The packer jobs [here](https://jenkins.hyperledger.org/view/ci-management/)  run automatically to create new images. Once the image is built, we usually test it once in sandbox and roll it out.

Feel free to drop by https://chat.hyperledger.org/channel/ci-pipeline to discuss

**Comment by jt-nti** [comment]
A few comments/questions from @dselman on the community call today (yell if I've got these wrong/missed any):

- What kind of service level/build performance can we expect from Jenkins? _I think we'd be sharing a pool of build slaves with all Hyperledger projects so not sure how this compares with the five dedicated build engines we currently get on Travis._

- How will the builds be migrate from Travis to Jenkins. I assume both will have to run in parallel? _Yes, and my current thoughts are that it would be good to keep Travis builds running even after the migration since this gives people the option to easily run Travis builds in their own fork._

- Why a self-managed Jenkins infrastructure?

**Comment by ryjones** [comment]
> - What kind of service level/build performance can we expect from Jenkins?

> _I think we'd be sharing a pool of build slaves with all Hyperledger projects so not sure how this compares with the five dedicated build engines we currently get on Travis._

Jenkins dynamically allocates up to 45 use-once build minions, across all projects. If we hit that limit we will ask the TSC to allocate more money. We've never hit the limit.

> - How will the builds be migrate from Travis to Jenkins. I assume both will have to run in parallel?
> _Yes, and my current thoughts are that it would be good to keep Travis builds running even after the migration since this gives people the option to easily run Travis builds in their own fork._

No opinion

> - Why a self-managed Jenkins infrastructure?

Could you expand on this?

**Comment by tykeal** [comment]
> To support Windows, we would need to use something like Azure. Again, the expense is pretty high. 

To be precise, this is what is required for us to build on Windows 10. The cloud being used for Jenkins supports Windows Server 2016, Standard edition and we have some Packer templates being developed by another project supported by LF to bring up instances on it.

**Comment by jt-nti** [comment]
@tykeal I thought Fabric supported Windows so would be a bit surprised if Windows wasn't supported at all. Composer doesn't yet work on Windows but is something that gets requested, so it'd be nice to have available at some point.

@ryjones re: Jenkins, it just seemed that self hosting a build environment wasn't the way projects would typically go these days. I guess that platform coverage is the biggest issue, which is a pity because Travis is pretty good otherwise.

@jasontclark is interested in getting familiar with the Composer CI process and offer his CI expertise

...on a related note, I've created a [Builds wiki page](https://github.com/hyperledger/composer/wiki/Builds) to collect information on the existing builds (mostly from @mbwhite I'm guessing!) and sketch out any proposals.

**Comment by sstone1** [comment]
@jt-nti in my experience, Fabric _claim_ to support Windows, but then somebody actually tries to use it on Windows and discovers it doesn't work - because they have no automated CI on Windows - and then somebody has to go and fix it long after the breaking change has gone in. The latest example of this was the Fabric Node.js SDK adding a dependency on a native module that refused to install on Windows.

**Comment by ryjones** [comment]
@jt-nti when Fabric was rolling out, Windows was not available to us at our service provider. It is now. I had forgotten that Azure was needed for client versions of Windows, not server versions.

@sstone1 Arnaud was pretty good at the start at keeping Windows BVT issues at the forefront. Fabric has had a similar issue with vagrant versus docker-compose, etc. It's a vicious cycle: low dev attention leads to low user adoption.
